cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2015 Tigera Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	"github.com/projectcalico/libcalico-go/lib/logutils"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	"github.com/projectcalico/cni-plugin/internal/pkg/utils"
+)
+
+const (
+	ipAllocPath = "/var/lib/cni/networks/k8s-pod-network"
+)
+
+var (
+	binariesToDisable = []string{
+		"/host/opt/cni/bin/calico",
+		"/host/opt/cni/bin/host-local",
+	}
+)
+
+type Accessor interface {
+	Backend() bapi.Client
+}
+
+func main() {
+	log.Info("migrating from host-local to calico-ipam...")
+	ctxt := context.Background()
+
+	// nodename associates IPs to this node.
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	logCtxt := log.WithField("node", nodename)
+
+	// calicoClient makes IPAM calls.
+	cfg, err := apiconfig.LoadClientConfig("")
+	if err != nil {
+		logCtxt.Fatal("failed to load api client config")
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	calicoClient, err := client.New(*cfg)
+	if err != nil {
+		logCtxt.Fatal("failed to initialize api client")
+	}
+
+	// Execute migration
+	if err := Migrate(ctxt, calicoClient, nodename); err != nil {
+		logCtxt.WithError(err).Fatal("failed to migrate ipam")
+	}
+}
+
+func Migrate(ctxt context.Context, c client.Interface, nodename string) error {
+	// Set up logging formatting.
```

Comment:
These should be at the beginning of execution, so that all logs get this formatting.

As it is, we're emitting some logs before these lines get executed.